src/coreclr/jit/inlinepolicy.cpp

```diff
@@ -711,7 +711,23 @@ void DefaultPolicy::NoteInt(InlineObservation obs, int value)
             }
             else if (!m_IsForceInline && (basicBlockCount > MAX_BASIC_BLOCKS))
             {
-                SetNever(InlineObservation::CALLEE_TOO_MANY_BASIC_BLOCKS);
+                unsigned basicBlockLimit = MAX_BASIC_BLOCKS;
+                if (!m_RootCompiler->opts.IsReadyToRun() && !JitConfig.JitConservativeInliner())
+                {
+                    // Allow to handle more than MAX_BASIC_BLOCKS blocks if we noted foldable branches, throw blocks
+                    // or have Profile Data.
+                    basicBlockLimit += m_FoldableBranch + m_ThrowBlock;
+                    if (m_HasProfile)
+                    {
+                        basicBlockLimit += static_cast<unsigned>(min(m_ProfileFrequency, 1.0) * 3.0);
+                    }
+                    basicBlockLimit = min(12, basicBlockCount);
+                }
+                if (basicBlockCount > basicBlockLimit)
+                {
+                    JITDUMP("basicBlockCount=%d > basicBlockLimit=%d\n", basicBlockCount, basicBlockLimit);
+                    SetNever(InlineObservation::CALLEE_TOO_MANY_BASIC_BLOCKS);
+                }
```

Comment:
I never really understood why we had this limit in the first place. And with the new logic it looks even more arbitrary than before. So I would strongly consider not having a block limit for the "new" inliner.

Ideally I'd like to see this sort of behavior encapsulated via a new policy rather than a check like you're doing above -- that is we'd always observe the number of blocks but this new policy would simply ignore the observation.

Reply:
Fixed to ignore this limit in `ExtendedDefaultPolicy` - will check perf impact.